### PR TITLE
Add start BlobStore API on server side

### PR DESCRIPTION
**Pull request by jsjtzyy**
Add start BlobStore feature into handleBlobStoreControlRequest method.
Also, changes have been made in StorageManager, DiskManager and
CompactionManager to properly start and control compaction on blobstore.
Deregister methods are added in StoreMetrics to remove associated
metrics when blobstore is shutdown.

### CODE REVIEW

File: ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java [hunk 1]
```diff
@@ -779,17 +779,40 @@ private AdminResponse handleBlobStoreControlRequest(DataInputStream requestStrea
...
+    PartitionId partitionId = blobStoreControlAdminRequest.getPartitionId();
+    if (partitionId != null) {
+      error = validateRequest(partitionId, RequestOrResponseType.AdminRequest);
+      if (error.equals(ServerErrorCode.No_Error)) {
+        if (blobStoreControlAdminRequest.shouldEnable()) {
+          // start BlobStore properly
+          if (storageManager.startBlobStore(partitionId)) {
             Collection<PartitionId> partitionIds = Collections.singletonList(partitionId);
-            if (storageManager.disableCompactionForBlobStore(partitionId)) {
+            controlRequestForPartitions(RequestOrResponseType.GetRequest, partitionIds, true);
```
Comment (vgkholla) [inline ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java]: I think we should modify this function to take in an `EnumSet`. That way we can compress the four lines into one and make this more concise (and some code in the stop part also)

File: ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java [hunk 2]
```diff
@@ -779,17 +779,40 @@ private AdminResponse handleBlobStoreControlRequest(DataInputStream requestStrea
...
+            controlRequestForPartitions(RequestOrResponseType.ReplicaMetadataRequest, partitionIds, true);
+            controlRequestForPartitions(RequestOrResponseType.PutRequest, partitionIds, true);
+            controlRequestForPartitions(RequestOrResponseType.DeleteRequest, partitionIds, true);
+            if (replicationManager.controlReplicationForPartitions(partitionIds, Collections.<String>emptyList(),
+                true)) {
+              if (storageManager.controlCompactionForBlobStore(partitionId, true)) {
+                error = ServerErrorCode.No_Error;
+                logger.info("Compaction is successfully enabled for partition: {}", partitionId);
```
Comment (vgkholla) [inline ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java]: Would prefer a message saying store started and functional for partition ID

File: ambry-store/src/main/java/com.github.ambry.store/CompactionManager.java [hunk 3]
```diff
@@ -314,11 +315,18 @@ boolean scheduleNextForCompaction(BlobStore store) {
...
      * Disable the compaction on given blobstore
-     * @param store the {@link BlobStore} on which the compaction is disabled.
+     * @param store the {@link BlobStore} on which the compaction is enabled or disabled.
+     * @param enabled whether to enable ({@code true}) or disable.
      * @return {@code true} if the disabling was successful. {@code false} if not.
      */
-    boolean disableCompactionForBlobStore(BlobStore store) {
-      storesDisabledCompaction.add(store);
+    boolean controlCompactionForBlobStore(BlobStore store, boolean enabled) {
+      if (enabled) {
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/CompactionManager.java]: `if (enabled && storesDisabledCompaction.contains(store)`
`else if (!enabled)`

Why do you need the `contains` check?

File: ambry-store/src/main/java/com.github.ambry.store/StoreMetrics.java [hunk 4]
```diff
@@ -176,6 +176,13 @@ void initializeIndexGauges(String storeId, final PersistentIndex index, final lo
     registry.register(MetricRegistry.name(Log.class, prefix + "CurrentSegmentCount"), currentSegmentCount);
   }
 
+  boolean deregisterIndexGauges(String storeId) {
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/StoreMetrics.java]: add documentation

Comment (vgkholla) [reply ambry-store/src/main/java/com.github.ambry.store/StoreMetrics.java]: this function can be `private`

File: ambry-store/src/main/java/com.github.ambry.store/StoreMetrics.java [hunk 5]
```diff
@@ -195,10 +202,27 @@ void initializeHardDeleteMetric(String storeId, final HardDeleter hardDeleter, f
     registry.register(MetricRegistry.name(PersistentIndex.class, prefix + "HardDeleteCaughtUp"), hardDeleteCaughtUp);
   }
 
+  boolean deregisterHardDeleteMetric(String storeId) {
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/StoreMetrics.java]: this function can be `private`

File: ambry-store/src/main/java/com.github.ambry.store/StoreMetrics.java [hunk 6]
```diff
@@ -195,10 +202,27 @@ void initializeHardDeleteMetric(String storeId, final HardDeleter hardDeleter, f
...
+  }
+
   void initializeCompactorGauges(String storeId, final AtomicBoolean compactionInProgress) {
     String prefix = storeId + SEPERATOR;
     Gauge<Long> compactionInProgressGauge = () -> compactionInProgress.get() ? 1L : 0L;
     registry.register(MetricRegistry.name(BlobStoreCompactor.class, prefix + "CompactionInProgress"),
         compactionInProgressGauge);
   }
+
+  boolean deregisterCompactorGauges(String storeId) {
```
Comment (vgkholla) [inline ambry-store/src/main/java/com.github.ambry.store/StoreMetrics.java]: this function can be `private`

File: ambry-store/src/main/java/com.github.ambry.store/CompactionManager.java (hunk 3 above)
Comment (jsjtzyy) [reply ambry-store/src/main/java/com.github.ambry.store/CompactionManager.java]: Unnecessary `contains` check is removed.

File: ambry-store/src/test/java/com.github.ambry.store/CompactionManagerTest.java [hunk 7]
```diff
@@ -347,9 +347,13 @@ public void testDisableCompactionForBlobStore() throws Exception {
     compactionManager = new CompactionManager(MOUNT_PATH, config, Collections.singleton(blobStore), metrics, time);
     compactionManager.enable();
     assertTrue("Disable compaction on given BlobStore should succeed",
-        compactionManager.disableCompactionForBlobStore(blobStore));
+        compactionManager.controlCompactionForBlobStore(blobStore, false));
     assertFalse("BlobStore should not be scheduled after compaction is disabled on it",
         compactionManager.scheduleNextForCompaction(blobStore));
+    assertTrue("Enable compaction on given BlobStore should succeed",
```
[the remark this answers is not in the document]
Comment (jsjtzyy) [reply ambry-store/src/test/java/com.github.ambry.store/CompactionManagerTest.java]: Periodic compaction test has been added in `testCompactionWithMisbehavingStores()`.

File: ambry-server/src/integration-test/java/com.github.ambry.server/ServerTestUtil.java [hunk 8]
```diff
@@ -344,6 +350,106 @@ protected static void endToEndTest(Port targetPort, String routerDatacenter, Str
       stream = channel.receive().getInputStream();
       GetResponse resp5 = GetResponse.readFrom(new DataInputStream(stream), clusterMap);
       assertEquals(ServerErrorCode.Blob_Not_Found, resp5.getPartitionResponseInfoList().get(0).getErrorCode());
+
+      System.out.println("Begin to stop a BlobStore");
+      // stop the store via AdminRequest
+      AdminRequest adminRequest =
+          new AdminRequest(AdminRequestOrResponseType.BlobStoreControl, partitionIds.get(0), 1, "clientid2");
+      BlobStoreControlAdminRequest controlRequest = new BlobStoreControlAdminRequest((short) 3, false, adminRequest);
```
Comment (vgkholla) [inline ambry-server/src/integration-test/java/com.github.ambry.server/ServerTestUtil.java]: will 0 work here (instead of 3)? - I want to avoid unpredictable failures due to thread scheduling

File: ambry-server/src/integration-test/java/com.github.ambry.server/ServerTestUtil.java [hunk 9]
```diff
@@ -344,6 +350,106 @@ protected static void endToEndTest(Port targetPort, String routerDatacenter, Str
...
+
+      System.out.println("Begin to restart the BlobStore");
+      // start the store via AdminRequest
+      adminRequest = new AdminRequest(AdminRequestOrResponseType.BlobStoreControl, partitionIds.get(0), 1, "clientid2");
+      controlRequest = new BlobStoreControlAdminRequest((short) 0, true, adminRequest);
+      channel.send(controlRequest);
+      stream = channel.receive().getInputStream();
+      adminResponse = AdminResponse.readFrom(new DataInputStream(stream));
+      assertEquals("Start store admin request should succeed", ServerErrorCode.No_Error, adminResponse.getError());
+      Thread.sleep(200);
```
Comment (vgkholla) [inline ambry-server/src/integration-test/java/com.github.ambry.server/ServerTestUtil.java]: why the sleep?

File: ambry-store/src/test/java/com.github.ambry.store/StorageManagerTest.java [hunk 10]
```diff
@@ -163,7 +230,7 @@ public void shutdownBlobStoreTest() throws Exception {
     List<? extends ReplicaId> invalidPartitionReplicas = invalidPartition.getReplicaIds();
     StorageManager storageManager = createStorageManager(replicas, metricRegistry);
     storageManager.start();
-    for (int i = 0; i < replicas.size() - 1; i++) {
+    for (int i = 1; i < replicas.size() - 1; i++) {
```
Comment (vgkholla) [inline ambry-store/src/test/java/com.github.ambry.store/StorageManagerTest.java]: why the change from 0 to 1?

File: ambry-server/src/test/java/com.github.ambry.server/AmbryRequestsTest.java [hunk 11]
```diff
@@ -356,38 +356,98 @@ public void stopBlobStoreSuccessTest() throws InterruptedException, IOException
...
+    // start BlobStore
+    adminRequest = new AdminRequest(AdminRequestOrResponseType.BlobStoreControl, id, correlationId, clientId);
+    BlobStoreControlAdminRequest startBlobStoreAdminRequest =
+        new BlobStoreControlAdminRequest(numReplicasCaughtUpPerPartition, true, adminRequest);
+    response = sendRequestGetResponse(startBlobStoreAdminRequest, ServerErrorCode.No_Error);
+    assertTrue("Response not of type AdminResponse", response instanceof AdminResponse);
+    // verify APIs are called in the process of starting BlobStore
+    assertEquals("Partition started not as expected", id, storageManager.startedPartitionId);
+    assertEquals("Replication on given BlobStore should be enabled", true, replicationManager.enableVal);
+    assertEquals("Partition enabled for compaction not as expected", id,
```
Comment (vgkholla) [inline ambry-server/src/test/java/com.github.ambry.server/AmbryRequestsTest.java]: shouldn't you check whether the enable value is `true` or `false`?

File: ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java [hunk 12]
```diff
@@ -109,6 +109,7 @@
   private final StoreKeyFactory storeKeyFactory;
   private final ConcurrentHashMap<RequestOrResponseType, Set<PartitionId>> requestsDisableInfo =
       new ConcurrentHashMap<>();
+  private final Set<String> replicaMetadataRequestDisabledStores = ConcurrentHashMap.newKeySet();
```
Comment (vgkholla) [inline ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java]: will chat with you offline about why this is required

File: ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java [hunk 13]
```diff
@@ -921,33 +959,42 @@ private void sendGetResponse(RequestResponseChannel requestResponseChannel, GetR
    * Check that the provided partition is valid, on the disk, and can be written to.
    * @param partition the partition to validate.
    * @param requestType the {@link RequestOrResponseType} being validated.
+   * @param skipConditionsCheck whether to skip ({@code true}) or perform conditions check.
```
Comment (vgkholla) [inline ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java]: variable needs to be named better to describe what checks are being skipped (and doc too)

File: ambry-server/src/integration-test/java/com.github.ambry.server/ServerTestUtil.java [hunk 14]
```diff
@@ -344,6 +351,108 @@ protected static void endToEndTest(Port targetPort, String routerDatacenter, Str
...
+
+      System.out.println("Begin to stop a BlobStore");
+      // stop the store via AdminRequest
+      AdminRequest adminRequest =
+          new AdminRequest(AdminRequestOrResponseType.BlobStoreControl, partitionIds.get(0), 1, "clientid2");
+      BlobStoreControlAdminRequest controlRequest = new BlobStoreControlAdminRequest((short) 0, false, adminRequest);
+      channel.send(controlRequest);
+      stream = channel.receive().getInputStream();
+      AdminResponse adminResponse = AdminResponse.readFrom(new DataInputStream(stream));
+      if (adminResponse.getError() == ServerErrorCode.Retry_After_Backoff) {
```
Comment (vgkholla) [inline ambry-server/src/integration-test/java/com.github.ambry.server/ServerTestUtil.java]: since we do 0 now, this should never happen right?

File: ambry-store/src/main/java/com.github.ambry.store/CompactionManager.java [hunk 15]
```diff
@@ -98,6 +98,7 @@ public void uncaughtException(Thread t, Throwable e) {
           logger.error("Thread {} threw exception", t, e);
         }
       });
+      compactionExecutor.enable();
```
[the remark this answers is not in the document]
Comment (jsjtzyy) [reply ambry-store/src/main/java/com.github.ambry.store/CompactionManager.java]: Fix this by setting enabled to false in CompactionExecutor.

File: ambry-store/src/test/java/com.github.ambry.store/StorageManagerTest.java (hunk 10 above)
Comment (jsjtzyy) [reply ambry-store/src/test/java/com.github.ambry.store/StorageManagerTest.java]: First store is used to test a special case where we try to stop a store when diskManager is not running.

File: ambry-server/src/test/java/com.github.ambry.server/AmbryRequestsTest.java [hunk 16]
```diff
@@ -922,37 +993,53 @@ public void shutdown() throws StoreException {
...
     /**
-     * The {@link PartitionId} that was provided in the call to {@link #disableCompactionForBlobStore(PartitionId)}
+     * The {@link PartitionId} that was provided in the call to {@link #controlCompactionForBlobStore(PartitionId, boolean)}
+     */
+    PartitionId compactionControlledPartitionId = null;
+    /**
+     * The {@link boolean} that was provided in the call to {@link #controlCompactionForBlobStore(PartitionId, boolean)}
      */
-    PartitionId compactionDisabledPartitionId = null;
+    Boolean enableVal = null;
```
Comment (vgkholla) [inline ambry-server/src/test/java/com.github.ambry.server/AmbryRequestsTest.java]: `compactionEnableVal`

File: ambry-server/src/integration-test/java/com.github.ambry.server/ServerTestUtil.java [hunk 17]
```diff
@@ -1056,8 +1165,122 @@ public void onCompletion(String result, Exception exception) {
...
+    Port port =
+        new Port(portType == PortType.PLAINTEXT ? dataNodeId.getPort() : dataNodeId.getSSLPort(), portType);
+    ConnectedChannel channel = connectionPool.checkOutConnection("localhost", port, 10000);
+    ArrayList<BlobId> ids = new ArrayList<BlobId>();
+    ArrayList<PartitionRequestInfo> partitionRequestInfoList = new ArrayList<PartitionRequestInfo>();
+    Payload payload1 = payloadQueue.poll(1000, TimeUnit.MILLISECONDS);
+    BlobId blobId1 = new BlobId(payload1.blobId, clusterMap);
+    ids.add(blobId1);
+    PartitionId partitionId = ids.get(0).getPartition();
+    byte[] usermetadata = new byte[1000];
```
Comment (vgkholla) [inline ambry-server/src/integration-test/java/com.github.ambry.server/ServerTestUtil.java]: create all this when you need it - code as currently written is a bit confusing

File: ambry-store/src/test/java/com.github.ambry.store/CompactionManagerTest.java [hunk 18]
```diff
@@ -20,6 +20,7 @@
 import com.github.ambry.utils.SystemTime;
 import com.github.ambry.utils.TestUtils;
 import com.github.ambry.utils.Time;
+import com.sun.scenario.effect.impl.sw.sse.SSEBlend_SRC_OUTPeer;
```
Comment (vgkholla) [inline ambry-store/src/test/java/com.github.ambry.store/CompactionManagerTest.java]: what is this? maybe file needs formatting and optimizing of imports? (check other files too)

File: ambry-server/src/integration-test/java/com.github.ambry.server/ServerTestUtil.java [hunk 19]
```diff
@@ -1056,8 +1170,116 @@ public void onCompletion(String result, Exception exception) {
...
+    stream = channel.receive().getInputStream();
+    resp1 = GetResponse.readFrom(new DataInputStream(stream), clusterMap);
+    InputStream responseStream = resp1.getInputStream();
+    BlobAll blobAll = MessageFormatRecord.deserializeBlobAll(responseStream, blobIdFactory);
+    byte[] actualBlobData = new byte[(int) blobAll.getBlobData().getSize()];
+    blobAll.getBlobData().getStream().getByteBuffer().get(actualBlobData);
+    Assert.assertArrayEquals("Content mismatch", data, actualBlobData);
+
+    // delete a blob on a restarted store , which should succeed
+    deleteRequest = new DeleteRequest(1, "deleteClient", blobId1, System.currentTimeMillis());
```
Comment (vgkholla) [inline ambry-server/src/integration-test/java/com.github.ambry.server/ServerTestUtil.java]: change this to `blobId2` too?

File: ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java [hunk 20]
```diff
@@ -816,20 +849,21 @@ private AdminResponse handleBlobStoreControlRequest(DataInputStream requestStrea
...
         }
-      } else {
-        error = ServerErrorCode.Partition_Unknown;
-        logger.debug("The partition Id should not be null.");
       }
     } else {
-      error = ServerErrorCode.Bad_Request;
-      logger.debug("The number of replicas to catch up should not be less or equal to zero {}",
-          blobStoreControlAdminRequest.getNumReplicasCaughtUpPerPartition());
+      error = ServerErrorCode.Partition_Unknown;
```
Comment (pnarayanan) [inline ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java]: Absence of a partition altogether should remain as a bad request. Partition unknown really means that there is a partition in the request that is not recognized by this node.

File: ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java [hunk 21]
```diff
@@ -921,33 +955,38 @@ private void sendGetResponse(RequestResponseChannel requestResponseChannel, GetR
...
-    List<? extends ReplicaId> replicaIds = partition.getReplicaIds();
-    for (ReplicaId replica : replicaIds) {
-      if (replica.getDataNodeId().getHostname().equals(currentNode.getHostname())
-          && replica.getDataNodeId().getPort() == currentNode.getPort()) {
-        if (replica.getDiskId().getState() == HardwareState.UNAVAILABLE) {
+    if (!skipPartitionAndDiskAvailableCheck) {
+      // 2. check if partition exists on this node and that the store for this partition has been started
+      if (storageManager.getStore(partition) == null) {
+        if (partitionsInCurrentNode.contains(partition)) {
           metrics.diskUnavailableError.inc();
```
Comment (pnarayanan) [inline ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java]: As discussed, Disk_Unavailable is not appropriate here. Perhaps this should be `Partition_Unavailable` (so that a caller does not think that all partitions on the said disk are down).

File: ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java [hunk 22]
```diff
@@ -779,29 +782,59 @@ private AdminResponse handleBlobStoreControlRequest(DataInputStream requestStrea
...
+    if (partitionId != null) {
+      if (blobStoreControlAdminRequest.shouldEnable()) {
+        // start BlobStore properly
+        error = validateRequest(partitionId, RequestOrResponseType.AdminRequest, true);
         if (error.equals(ServerErrorCode.No_Error)) {
-          if (blobStoreControlAdminRequest.shouldEnable()) {
-            // TODO: start BlobStore properly
-            error = ServerErrorCode.Temporarily_Disabled;
-          } else {
+          if (storageManager.startBlobStore(partitionId)) {
```
Comment (pnarayanan) [inline ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java]: As in the case of shutting down, we should make sure that the `ClusterParticipant` is made aware of this, so it can synchronously update the status of the store/replica in Helix.

Comment (vgkholla) [reply ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java]: I think that we should reserve all clustermap interaction to another PR.

File: ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java (hunk 21 above)
Comment (jsjtzyy) [reply ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java]: Will introduce `Partition_Unavailable` error code in a new PR.

File: ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java (hunk 22 above)
Comment (jsjtzyy) [reply ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java]: Changes related to clustermap have been mentioned in issue #882 ,  which will be made in a new PR.
